pkg/provider/azure_local_services.go

```diff
@@ -218,3 +275,245 @@ func newBatchOperationResult(name string, success bool, err error) batchOperatio
+func getLocalServiceBackendPoolName(serviceName string, ipv6 bool) string {
+	serviceName = strings.ToLower(strings.Replace(serviceName, "/", "-", -1))
+	if ipv6 {
+		return fmt.Sprintf("%s-ipv6", serviceName)
+	}
+	return serviceName
+}
+
+// getBackendPoolNameForService determine the expected backend pool name
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolNameForService(service *v1.Service, clusterName string, ipv6 bool) string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return getBackendPoolName(clusterName, ipv6)
+	}
+	return getLocalServiceBackendPoolName(getServiceName(service), ipv6)
+}
+
+// getBackendPoolNamesForService determine the expected backend pool names
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolNamesForService(service *v1.Service, clusterName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return getBackendPoolNames(clusterName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: getLocalServiceBackendPoolName(getServiceName(service), false),
+		consts.IPVersionIPv6: getLocalServiceBackendPoolName(getServiceName(service), true),
+	}
+}
+
+// getBackendPoolIDsForService determine the expected backend pool IDs
+// by checking the external traffic policy of the service.
+func (az *Cloud) getBackendPoolIDsForService(service *v1.Service, clusterName, lbName string) map[bool]string {
+	if !isLocalService(service) || !az.useMultipleStandardLoadBalancers() {
+		return az.getBackendPoolIDs(clusterName, lbName)
+	}
+	return map[bool]string{
+		consts.IPVersionIPv4: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, false),
+		consts.IPVersionIPv6: az.getLocalServiceBackendPoolID(getServiceName(service), lbName, true),
+	}
+}
+
+// getLocalServiceBackendPoolID gets the ID of the backend pool of a local service.
+func (az *Cloud) getLocalServiceBackendPoolID(serviceName string, lbName string, ipv6 bool) string {
+	return az.getBackendPoolID(lbName, getLocalServiceBackendPoolName(serviceName, ipv6))
+}
+
+// localServiceOwnsBackendPool checks if a backend pool is owned by a local service.
+func localServiceOwnsBackendPool(serviceName, bpName string) bool {
+	prefix := strings.Replace(serviceName, "/", "-", -1)
+	return strings.HasPrefix(strings.ToLower(bpName), strings.ToLower(prefix))
+}
+
+type serviceInfo struct {
+	ipFamily string
+	lbName   string
+}
+
+func newServiceInfo(ipFamily, lbName string) *serviceInfo {
+	return &serviceInfo{
+		ipFamily: ipFamily,
+		lbName:   lbName,
+	}
+}
+
+// getLocalServiceEndpointsNodeNames gets the node names that host all endpoints of the local service.
+func (az *Cloud) getLocalServiceEndpointsNodeNames(service *v1.Service) (sets.Set[string], error) {
+	eps, err := az.KubeClient.DiscoveryV1().EndpointSlices(service.Namespace).List(context.Background(), metav1.ListOptions{})
```

Comment:
Sure, will add a cache.

Reply:
Added a cache, please check the latest commit. 
BTW, there is a problem theoretically: if the endpointslice doesn't finish updating when we ensure hosts in pool, there could be a discrepancy between the # of node IPs in the pool and the actual state. This is because the endpoint slice update may be slower than the pods update. After testing multiple times, I haven't found this behavior but I think this could happen.

Reply:
yes, that's a valid issue on large scale of workloads. Could we also trigger the backend pool updates on the endpointslice add/delete events?

Reply:
Synced offline. There is no need to support backend pool updates for add/delete events. The latency issue will be fixed in a patch version of v1.28.